pageserver/src/basebackup.rs

```diff
@@ -0,0 +1,197 @@
+use log::*;
+use tar::{Builder};
+use std::fmt;
+use std::io::Write;
+use walkdir::WalkDir;
+use regex::Regex;
+
+use crate::ZTimelineId;
+
+
+pub fn send_snapshot_tarball(write: &mut dyn Write, timelineid: ZTimelineId, snapshotlsn: u64) -> Result<(), std::io::Error> {
+    let mut ar = Builder::new(write);
+
+    let snappath = format!("timelines/{}/snapshots/{:016X}", timelineid, snapshotlsn);
+    let walpath = format!("timelines/{}/wal", timelineid);
+
+    debug!("sending tarball of snapshot in {}", snappath);
+    //ar.append_dir_all("", &snappath)?;
+
+    for entry in WalkDir::new(&snappath) {
+        let entry = entry?;
+        let fullpath = entry.path();
+        let relpath = entry.path().strip_prefix(&snappath).unwrap();
+
+        if relpath.to_str().unwrap() == "" {
+            continue;
+        }
+
+        if entry.file_type().is_dir() {
+            trace!("sending dir {} as {}", fullpath.display(), relpath.display());
+            ar.append_dir(relpath, fullpath)?;
+        } else if entry.file_type().is_symlink() {
+            error!("ignoring symlink in snapshot dir");
+        } else if entry.file_type().is_file() {
+
+            // Shared catalogs are exempt
+            if relpath.starts_with("global/") {
+                trace!("sending shared catalog {}", relpath.display());
+                ar.append_path_with_name(fullpath, relpath)?;
+            } else if !is_rel_file_path(relpath.to_str().unwrap()) {
+                trace!("sending {}", relpath.display());
+                ar.append_path_with_name(fullpath, relpath)?;
+            } else {
+                trace!("not sending {}", relpath.display());
+                // FIXME: send all files for now
+                ar.append_path_with_name(fullpath, relpath)?;
+            }
+        } else {
+            error!("unknown file type: {}", fullpath.display());
+        }
+    }
+
+    // FIXME: also send all the WAL
+    for entry in std::fs::read_dir(&walpath)? {
+        let entry = entry?;
+        let fullpath = &entry.path();
+        let relpath = fullpath.strip_prefix(&walpath).unwrap();
+
+        if !entry.path().is_file() {
+            continue;
+        }
+
+        let archive_fname = relpath.to_str().unwrap().clone();
```

Comment:
clippy says:
```
error: using `clone` on a double-reference; this will copy the reference of type `&str` instead of cloning the inner type
  --> pageserver/src/basebackup.rs:69:29
   |
69 |         let archive_fname = relpath.to_str().unwrap().clone();
   |                             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: `#[deny(clippy::clone_double_ref)]` on by default
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#clone_double_ref
```
The `.clone()` can be safely deleted.

There are several instances of this error in this PR.